{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ace48285-6a5e-4251-95a3-4357fb0b5e6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: tqdm in c:\\users\\ajitw\\anaconda3\\envs\\llmenv\\lib\\site-packages (4.66.2)\n",
      "Requirement already satisfied: colorama in c:\\users\\ajitw\\anaconda3\\envs\\llmenv\\lib\\site-packages (from tqdm) (0.4.6)\n"
     ]
    }
   ],
   "source": [
    "# Dependancy\n",
    "# !pip install transformers\n",
    "# !pip install PyMuPDF\n",
    "# !pip install sentencepiece\n",
    "# !pip install --upgrade tensorflow\n",
    "# install cuda(11.8) before pythorch \n",
    "# !pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118\n",
    "# !pip install tqdm "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "16f6ec1a-a2d4-4444-a033-34cf82ba2931",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.\n"
     ]
    }
   ],
   "source": [
    "from transformers import T5ForConditionalGeneration, T5Tokenizer\n",
    "import tqdm as notebook_tqdm\n",
    "import fitz \n",
    "\n",
    "model_name = \"google/flan-t5-small\"\n",
    "tokenizer = T5Tokenizer.from_pretrained(model_name)\n",
    "model = T5ForConditionalGeneration.from_pretrained(model_name)\n",
    "text = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dd1c3fb-8996-4519-a03b-e134a05bc2f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import fitz \n",
    "doc = fitz.open('sample.pdf') \n",
    "text = \"\" \n",
    "for page in doc: \n",
    "   text+=page.get_text() \n",
    "print(text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ab80ea7b-9680-4046-bf98-880528656119",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Extracts text from a PDF file.\n",
    "\"\"\"\n",
    "pdf_file_path = \"The_Human_Genome_An_Introduction.pdf\"\n",
    "# pdf_text = extract_text_from_pdf(pdf_file_path)\n",
    "\n",
    "text = \"\"\n",
    "with fitz.open(pdf_file_path) as pdf_document:\n",
    "    num_pages = pdf_document.page_count\n",
    "    for page_num in range(num_pages):\n",
    "        page = pdf_document[page_num]\n",
    "        text += page.get_text()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1ddaba81-e1d1-4395-95f6-4821537b0843",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'See discussions, stats, and author profiles for this publication at: https://www.researchgate.net/publication/12169019\\nThe Human Genome: An Introduction\\nArticle\\xa0\\xa0in\\xa0\\xa0The Oncologist · February 2001\\nDOI: 10.1634/theoncologist.2001-0100\\xa0·\\xa0Source: PubMed\\nCITATIONS\\n10\\nREADS\\n17,474\\n4 authors, including:\\nJeroen Aerssens\\nADQ\\n287 PUBLICATIONS\\xa0\\xa0\\xa07,410 CITATIONS\\xa0\\xa0\\xa0\\nSEE PROFILE\\nMartin Armstrong\\nUCB\\n63 PUBLICATIONS\\xa0\\xa0\\xa03,438 CITATIONS\\xa0\\xa0\\xa0\\nSEE PROFILE\\nRon Gilissen\\nJohnson & Johnson\\n56 PUBLICATIONS\\xa0\\xa0\\xa02,501 CITATIONS\\xa0\\xa0\\xa0\\nSEE PROFILE\\nAll content following this page was uploaded by Ron Gilissen on 01 June 2014.\\nThe user has requested enhancement of the downloaded file.\\nThe Human Genome: An Introduction\\nJEROEN AERSSENS, MARTIN ARMSTRONG, RON GILISSEN, NADINE COHEN\\nDepartment Pharmacogenomics, Janssen Research Foundation, Beerse, Belgium\\nThe Oncologist 2001;6:100-109\\nwww.TheOncologist.com\\nCorrespondence: Jeroen Aerssens, Ph.D., Department of Pharmacogenomics, Janssen Research Foundation, Turnhoutseweg 30,\\nBeerse B-2340, Belgium. Telephone: 32-14-606146; Fax: 32-14-607162; e-mail: jaerssen@janbe.jnj.com Received September\\n29, 2000; accepted for publication November 28, 2000. ©AlphaMed Press 1083-7159/2001/$5.00/0\\nThe\\nOncologist\\nFundamentals of Cancer Medicine\\nINTRODUCTION\\nDuring the past two decades, tremendous progress has\\nbeen made in genetics and genomics. Diseases that run in fam-\\nilies have been recognized for many centuries, but it was only\\nin the early 1980s that the first mutations in a gene responsible\\nfor a disease could be identified. Subsequently, numerous dis-\\ncoveries of disease-related mutations in other genes have been\\nfound, initially in rare single-gene disorders, but more recently\\nalso in common disorders such as Alzheimer’s disease and\\ncancer. Applications of this newly discovered information pro-\\nvide new opportunities for progression in medical science,\\nincluding the design of genetic tests to diagnose or predict\\n(subtypes of) diseases, the redefinition of diseases and the\\nunderstanding of their pathogenesis based on the molecular\\nmechanisms behind them, and the selection of new target mol-\\necules for drug discovery. The time has now come that such\\napplications will transfer further toward the day-to-day prac-\\ntice of the clinician and transform the practice of clinical med-\\nicine. For this to happen, an appropriate education and\\nunderstanding of the basic concepts of genetics and genomics\\nby clinicians is needed. This article aims to provide a general\\nintroduction of these concepts for clinicians not familiar with\\nthese fields. The list of references and the websites indicated in\\nthe manuscript should encourage the reader to get a broader\\nappreciation of this research area.\\nWHAT GENETICISTS ARE ALWAYS TALKING\\nABOUT: DNA\\nOur inherited information is encoded in a macromolecule\\ncalled “DNA” (deoxyribonucleic acid). Basically, our DNA is\\nlike a bar code—it encrypts information. The vast majority of\\nDNA molecules are stored within the nucleus of the cell and\\ncovered with proteins, together forming chromosomes. DNA\\ncompletely dissolves in aqueous solutions (and thus in its nat-\\nural environment). When precipitated in alcoholic solvents\\n(ethanol, isopropanol) during extraction procedures, the DNA\\nbecomes visible as a white viscous clot.\\nAll multicellular organisms, including humans, start\\ntheir life as a single cell (the fertilized egg), and almost all\\ncells of the organism developed from this single cell contain\\na full and identical copy of the DNA of this single cell. In\\nhumans, each somatic cell contains approximately 0.006\\nnanogram of DNA which harbors all the genetic information\\nneeded to develop and function normally. For an adult\\nhuman body, this makes a total of about 600 grams of DNA.\\nFor clinical genetic analyses, genomic DNA is usually iso-\\nlated from leukocytes, although identical DNA could be iso-\\nlated from virtually any other cell of an individual, as well.\\nFrom 5 to10 ml whole blood, approximately 100-200 micro-\\ngrams of DNA can be extracted, which is in most cases more\\nthan sufficient to perform genetic analyses.\\nTHE DNA IS IN THE CHROMOSOMES: TWO COPIES\\nOF A LIBRARY\\nThe DNA in our cells is divided over a constant number\\nof chromosomes (46 in humans), each of them with a specific\\nsize and form, as can be observed under the microscope using\\nspecific coloring techniques (e.g., Giemsa staining). Two sex\\nchromosomes (X and Y) determine the gender of an individ-\\nual (XX for females, XY for males); the other 44 so-called\\nautosomal chromosomes are not different in males and\\nfemales. Together, these 46 chromosomes comprise two\\nnearly identical copies of the whole genome: one copy of the\\ngenome is inherited from the father (via a set of 22 autosomal\\nAerssens, Armstrong, Gilissen et al.\\n101\\nchromosomes and an X or Y chromosome) and the other\\ncopy from the mother (via another set of 22 autosomal\\nchromosomes and one X chromosome).\\nThe autosomal chromosomes derived from the father and\\nmother are two-by-two homologous: they look similar under\\nthe microscope and comprise the same genes, or eventually,\\nvariants of the same genes. These chromosomes are num-\\nbered from 1 to 22, mainly based on size (1 being the largest\\nand 22 the smallest chromosome). Thus, each somatic cell\\ncontains two copies of each of these 22 chromosomes, and\\nthus two copies of each of the genes located on these chro-\\nmosomes. One could compare this with a library which con-\\ntains two copies of each book, although there might\\nsometimes be different editions of each specific book. In\\nwomen, the two X chromosomes are also homologous: one\\ncopy is inherited from their mother and the other copy from\\ntheir father. Men, on the contrary, have one X chromosome\\ninherited from their mother and one Y chromosome inherited\\nfrom their father. The Y chromosome is much smaller than\\nthe X chromosome and contains many fewer genes as well.\\nIn germ cells, only one copy of each homologous chro-\\nmosome is present (thus in total, 22 autosomal and one sex\\nchromosome). During reproduction, two germ cells (one\\negg cell and one sperm cell) combine their genetic infor-\\nmation so that the offspring will contain two copies of each\\nchromosome: one from the father and one from the mother\\n(Fig. 1). Thus, the gender of the offspring, determined by\\nthe combination of the sex chromosomes of the mother\\n(always X) and the father (X or Y) is completely dependent\\non whether the sperm cell contains an X or Y chromosome.\\nTHE SIZE AND STRUCTURE OF DNA—A DOUBLE\\nHELIX\\nFrom a structural point of view, the DNA looks like a long\\nchain of connected letters without any spaces or punctuation\\nmarks (Fig. 2A). The total physical length of all the DNA\\nchains in each of our cells is approximately two meters, with a\\ndiameter of 0.000002 mm. In order to write the DNA text, the\\nbody has at its disposal four different but related building\\nblocks (called nucleotides or, more precisely, deoxyribonu-\\ncleotides): A, C, G, and T, representing respectively adenine,\\ncytosine, guanine, and thymine. These nucleotides are con-\\nnected by a deoxyribose-phosphate backbone. Each phosphate\\nlinks the hydroxyl group on the 3′ carbon atom of a deoxyri-\\nbose of one nucleotide to the hydroxyl group on the 5′ carbon\\natom in the deoxyribose group of the adjacent nucleotide.\\nImportantly, all the information content encrypted within the\\nDNA is in the specific sequence of these nucleotides. The\\ninformation stored within the DNA code can be used for trans-\\nlation into functional activity (i.e., production of proteins) only\\nin one orientation on the backbone, namely the 5′-to-3′ direc-\\ntion. Therefore, nucleotide sequences are usually displayed\\nfrom the 5′-to-3′ end (from left to right).\\nAttached to this DNA strand is a second DNA strand\\nwhich is the exact complement of the first one. This is pos-\\nsible because of the complementary chemical structure of\\nthe DNA building blocks. Two kinds of base pairs, often\\nreferred to as complementary base pairs, exist in all DNA:\\nthe As on one strand always pair with Ts on the other strand\\n(via two hydrogen bonds) while Cs pair with Gs (via three\\nhydrogen bonds). Thus, if the sequence of one strand is\\nknown, the sequence of the com-\\nplementary strand can easily be\\nFigure 1. Schematic overview of the\\ninheritance of our genetic informa-\\ntion. Each individual has two copies of\\neach chromosome (each harboring\\none copy of the genes on the chromo-\\nsome). The germ cells comprise only\\none of these copies. Through combina-\\ntion of the genetic material from the\\nsperm cell and egg cell, the offspring\\ninherits one copy of each chromosome\\nfrom the father and one from the\\nmother. Assume a particular gene in\\nthe DNA which determines the pheno-\\ntype hair style, and for which two\\nvariants exist (A and G). The geno-\\ntype, which is the combination of the\\nvariants on the two inherited homolo-\\ngous chromosomes in an individual\\n(e.g., A/G), will determine the pheno-\\ntype. In the example, the G/G genotype is linked with straight hair, while individuals with the other genotypes (A/G and A/A) have curly hair\\n(indicating that the G variant, associated with straight hair, is a recessive characteristic).\\n102\\nThe Human Genome: An Introduction\\nderived. The two deoxyribose-phosphate backbones have\\nopposite 5′-to-3′ orientations and are wound around each\\nother to form a double-helix structure.\\nThe size of DNA molecules is noted as the number of base\\npairs (bp) or a multiple (1,000 bp = 1 kbp, 1,000 kbp = 1 Mbp).\\nOur complete library of genetic information is called “the\\nhuman genome,” and comprises somewhat more than three bil-\\nlion bp (3,000 Mbp), distributed over 22 autosomal chromo-\\nsomes (numbered from 1 to 22) and two sex chromosomes\\n(X and Y). A printed edition of this sequence would require\\napproximately one million printed pages with single-line spac-\\ning. The elucidation of this genomic DNA sequence is of\\nextreme interest, as it contains—in encrypted form—all the\\ninherited information needed to develop and direct the func-\\ntioning of the human body. Table 1 summarizes the dimensions\\nand information content of the human genome in numbers.\\nGENES IN THE GENOME\\nThe unit of information in the DNA is the gene, which is\\na stretch of DNA sequence that contains the code for the pro-\\nduction of a protein, which is a single piece of the whole\\nmachinery required by the cell to normally function in its\\nenvironment. More specifically, each gene comprises all the\\ndetailed instructions that determine the precise composition\\nof a specific protein, as well as the regulatory instructions\\nFigure 2. (A) The human genome sequence can be compared with a text lacking any spaces or punctuation marks. It is extremely difficult to\\nread the genomic text without an analysis tool; even special software programs which have been specifically written to identify meaningful sen-\\ntences (genes) only have a limited success rate. (B) This is because in between the words (the exons) of the text which form a meaningful sen-\\ntence (a gene), variable amounts of nonsense letters (the introns) are placed. In fact, more than 90% of the human sequence consists of text from\\nwhich the meaning is currently not understood. (C) Variation frequently occurs in the human genome (about one letter differs in every 1,000 let-\\nters between the genomic texts of two individuals). This might have consequences on the meaning of the sentence, or eventually make the sen-\\ntence unreadable. These variations in the DNA are called mutations or polymorphisms (depending on their frequency and on whether there is\\nor is not a direct link to the cause of a disease).\\nTable 1. The human genome in numbers\\n• 3,000,000,000 nucleotides in the human genome (estimated)\\n• 22 autosomal chromosomes and two sex chromosomes (X and Y)\\n• 46 chromosomes in each somatic cell (two copies of the whole \\ngenome)\\n• 30,000–120,000 genes in the human genome (estimated)\\n• 35,000 nucleotide sequences per gene at the genomic level, including \\nintronic sequences (on average)\\n• 1,500 nucleotides directly coding sequence per gene (on average)\\n• less than 5% of the human genome sequence directly encodes for \\nproteins\\n• four different nucleotides in the genome (adenine, cytosine, guanine, \\nthymine)\\n• three nucleotides comprised in a codon which encodes 1 amino acid\\n• one nucleotide difference between two unrelated individuals per \\n1,000 nucleotides sequence (on average)\\nAerssens, Armstrong, Gilissen et al.\\n103\\nthat determine when this specific protein will be produced\\nand in what quantity. The size of a gene at the genomic level\\ncan vary widely (usually between 10,000 and 150,000 bp).\\nAlthough most of the genomic DNA sequence is currently\\nknown, there is still a large debate ongoing on the number \\nof genes present: estimates of experts in the field vary\\nbetween 30,000 and 120,000, with an average around 60,000\\n(www.ensembl.org/genesweep.html). Very intriguingly, the\\nregions in the genomic DNA which encode for proteins\\naccount for about 150 million nucleotides, which is less than\\n5% of the complete human genome. Apart from some of the\\nDNA sequences which comprise instructions needed to reg-\\nulate the expression of the genes and specific instructions for\\nthe chromosomes to function correctly, the significance of\\nthe other 95% of the genome is at present largely unknown\\nand/or poorly understood. This latter part of the genome con-\\ntains large numbers of highly repeated DNA sequence fami-\\nlies. Two major types of repeat families can be distinguished:\\ntandemly repeated DNA and interspersed repetitive DNA.\\nTandemly repeated DNA families consist of long or short\\narrays of DNA repeat units, with the repeat being a simple or\\nmoderately complex sequence (size usually between 2 and\\n100 bp). Depending on the size of arrays of repeat units, this\\nis called satellite DNA (>100 kbp), minisatellite DNA (0.1-\\n20 kbp), or microsatellite DNA (<150 bp). Interspersed\\nrepetitive DNA consists of individual repeat units which are\\nnot clustered at a specific location on a chromosome, but are\\ndispersed at numerous locations. Among these are the SINEs\\n(short interspersed nuclear elements) and the LINEs (long\\ninterspersed nuclear elements). Well-known examples are\\nAlu repeats (SINE with full-length of 280 bp; approximately\\n1,000,000 copies in the human genome) and LINE-1 or L1\\nelement (LINE with full-length of 6.1 kbp; approximately\\n80,000 copies in the human genome).\\nTHE HUMAN GENOME PROJECT\\nIn 1987, a worldwide scientific effort called the Human\\nGenome Project was initiated to unravel the complete DNA\\nsequence of the human genome. Recently, a first draft \\ncovering 85%-90% of the complete human genome\\nsequence (3.12 billion bp) has been announced simultane-\\nously by scientists of the publicly funded Human Genome\\nProject (www.sanger.ac.uk/hgp; www.gene.ucl.ac.uk/huqo;\\nwww.nhgri.nih.gov; www.ncbi.nlm.nih.gov/genome/seq/)\\nand the private company Celera Genomics (Rockville, MD;\\nwww.celera.com). As a consequence of the strategies used to\\ndetermine the human genome sequence, experts anticipate\\nthat it will be another two years before the complete human\\ngenome sequence will be known with a confidence of more\\nthan 99.99% [1]. Although the complete sequence of the\\nwhole human genome will soon be known, it is expected that\\nit will take many more decades before all this information\\n(i.e., identification of all genes and their regulation, signifi-\\ncance of genetic variations, etc.) will be fully understood.\\nNevertheless, the scientific importance of the achievements\\nreached so far by the Human Genome Project can hardly be\\noverestimated and is at least of the same order of magnitude\\nas the Apollo lunar program.\\nFROM GENE TO PROTEIN\\nThe main role of the DNA in the cell is to permanently\\nstore and make available all the information needed to regulate\\neach of the activities in the cell. The production of proteins—\\nwhich are the functionally active molecules in the cell—takes\\nplace in the cytoplasm of the cell. Since the instructions for how\\nto make the proteins is within the DNA which is stored in the\\nnucleus, an intermediate molecule (messenger RNA, or\\nmRNA) is used to transfer this information from the nucleus to\\nthe cytoplasmic protein factory. As a matter of comparison,\\nimagine a library (the genomic DNA) which contains many\\nbooks (genes) on many different topics. A reader makes a copy\\n(the mRNA) of a specific book which contains the specific\\ninformation on how to make a cake and takes this to his home\\nas he is not allowed to make cake in the library. At home, the\\nperson can then make a cake (the protein) using all the required\\ningredients and supplies as described in the copied information.\\nWhen and how much of a gene should be expressed in a\\ncell is directed by specific proteins (transcription factors) which\\nare present in the nucleus and which can interact in a stimula-\\ntory or inhibitory manner with regulatory sequences in the\\nDNA flanking the coding part of the gene. When this fine-tuned\\nregulation mechanism indicates that additional copies of the\\ngene should be expressed, an enzyme in the nucleus (RNA\\npolymerase) transcribes the genetic information from the DNA\\ntemplate into an RNA (ribonucleic acid) copy. The structure of\\nRNA is similar to a single-strand DNA molecule, although\\nthymine (T) is replaced by uracil (U). Because the protein-cod-\\ning information in the DNA is interrupted by irrelevant\\nsequences (called introns), the RNA must be further edited\\n(spliced) to remove these intron sequences and join the coding\\nsequences (called exons) (Fig. 2B). In some genes, a choice\\nbetween several alternative exons is being made during this\\nsplicing process, which will result in different proteins. The\\nRNA molecule that results from transcription and splicing is\\ncalled messenger RNA (mRNA). This mRNA (on average\\n1,500 bp) is transported to the cytoplasm where it is used as a\\ntemplate for the generation of a protein. Thus, the mRNA is\\nthreaded through ribosomes as a tape is threaded through the\\nhead of a tape player in order to decode the information and\\nassemble the amino acids into chains. For the decoding, each\\nsubsequent group (called a “codon”) of three nucleotides on the\\nmRNA specifies a new amino acid. Mostly starting from a \\n104\\nThe Human Genome: An Introduction\\nso-called start codon with the sequence “ATG” (which encodes\\na methionine), each adjacent codon on the mRNA specifies the\\nnext amino acid to be linked to the growing protein chain. After\\ncompletion of this translation process, additional modifications\\nare made to the protein (e.g., phosphorylation, glycosylation),\\nresulting in a mature and functional protein.\\nIn summary, the properties of each protein depend on the\\nsequence of the amino acids used to construct it, and this\\nsequence in turn is determined directly by the nucleotide\\nsequence of the mRNA, which in turn is an (edited) copy of\\nthe genomic DNA sequence (Fig. 3). It should be noted that,\\nalthough generally the information for making any single\\nprotein is always encoded by a single gene, one gene may (as\\na result of differential splicing) carry the information needed\\nto make several (usually related) proteins.\\nGENE EXPRESSION IN THE CELL: WHICH GENES\\nAND IN WHAT AMOUNT\\nAs indicated above, the DNA content is identical in each\\ncell or tissue type of the body; however, not all our cells are\\nidentical in terms of structure, function, or behavior. What\\nmakes them different is the pattern of genes which are\\nexpressed and translated into proteins during the life cycle of\\nthe cells. Some cell types express many genes (e.g., in brain\\ncells approximately 30,000 genes are expressed), while in\\nothers a large number of the genes are transcriptionally \\ninactive (e.g., in red blood cells only 30 genes are expressed).\\nApart from an overall switching of the expression of specific\\ngenes from “on” to “off” (or vice versa), fine-tuning of the\\nexpression level of specific genes might also occur. Changes\\nin the level of expression may be the result of a disease or\\nmay eventually lead to a disease. Therefore, there is an enor-\\nmous scientific interest in studying and comparing the level\\nof expression of genes, i.e., gene expression in disease status\\nversus in healthy controls.\\nAnalysis of mRNA samples is very useful, as these con-\\ntain only the transcribed sequences of the human genome\\n(and thus the genes). Therefore, several research groups and\\nbiotech companies have cloned and analyzed large libraries\\nfrom mRNA sequences. For example, the mRNA extracted\\nfrom a brain sample contains copies of thousands of tran-\\nscribed genes which might be of interest. Technically, in\\norder to clone the transcribed genes, the mRNA molecules\\nfirst need to be converted into double-strand molecules. This\\ncan be done by adding the complement nucleotides on a sec-\\nond DNA strand (a process called “reverse transcription”),\\nresulting in double-stranded DNA molecules which contain\\nonly the exon sequences of the genes (but not the intron\\nsequences). These are called cDNA molecules (copy DNA),\\nand contain the open reading frame of the gene which can\\neasily be converted into the amino acid sequence of the\\nresulting protein. In large projects, several thousands of these\\nFigure 3. Schematic overview of how the information comprised within the genetic code is being used to synthesize the proteins. This\\ninvolves the processes of transcription from DNA into RNA, RNA splicing to form mRNA, transport of the mRNA from the nucleus to the cyto-\\nplasm, translation into a chain of amino acids, and finally post-translational modifications and folding of the synthesized protein. Note that at\\nboth the 5′ and 3′ ends of the coding region in the exonic sequence, an untranslated region (UTR) is also transcribed and spliced into mRNA\\n(respectively 5′-UTR and 3′-UTR regions).\\nAerssens, Armstrong, Gilissen et al.\\n105\\ncDNA clones have been partially sequenced and have\\nrevealed previously unknown fragments of expressed genes\\n(often called ESTs, expressed sequence tagged sites).\\nComparison of databases of EST sequences might eventu-\\nally also reveal new information on tissue-specific expression\\nof some genes. In the laboratory, the evaluation of gene\\nexpression levels in tissue samples can now also be evaluated\\nsimultaneously in thousands of genes, thanks to the enormous\\nprogression in the development of microarray technology\\n(more popularly, “DNA chip” technology) during the last few\\nyears. Today, this technology enables scientists to simultane-\\nously compare the expression levels of several thousand genes\\nin a single experiment, on a surface smaller than a stamp. This\\ntechnology is based on the hybridization of RNA samples\\n(e.g., extracted from diseased and healthy tissue) on glass\\nslides (DNA chips) containing DNA molecules with the spe-\\ncific sequences of thousands of different genes. The inten-\\nsity of the hybridization signals, which are a measure of the\\nexpression levels of the different genes, can be evaluated\\nusing powerful software [2].\\nVARIATIONS IN THE GENOME—THE BASIS OF\\nHUMAN DIVERSITY\\nWhen the DNA sequence of a gene is identified in dif-\\nferent individuals from the population, some differences in\\nthe nucleotide sequence are often detected (Fig. 2C). The\\ninformation content of DNA can be altered dramatically by\\nsuch variations in the nucleotide sequence, especially if\\nthese differences are located in protein-coding or regulatory\\nsequences. The consequence of such variations might lead\\nto the insertion of a different amino acid on a specific posi-\\ntion in the protein, or to a different level of expression of a\\nprotein. Variations located in the intronic regions of genes or\\noutside the genes will usually have fewer consequences. The\\ndifferent forms of a genetic variation are called the “alleles”\\nof the variation. Frequently occurring variations are often\\ncalled “polymorphisms,” while more rare variations (with\\nallele frequency below 1%) and variations with a direct rela-\\ntionship to a disease are often called “mutations” (although\\nthese definitions are arbitrary). Genetic variations can\\ninvolve only 1 bp (called single nucleotide polymorphism,\\nSNP), a few bp (e.g., di- and trinucleotide repeat polymor-\\nphisms), up to large stretches of DNA. Roughly, the varia-\\ntions can be divided into substitutions, insertions, deletions,\\namplifications, and translocations (Fig. 4).\\nThe major contributors to genetic variation, comprising\\nsome 80% of all known polymorphisms, are the single\\nnucleotide polymorphisms. An SNP located in the coding\\nregion of a gene is indicated as “cSNP.” It has been estimated\\nthat, on average, the DNA sequence of two unrelated individ-\\nuals differs in 0.1% (1 in 1,000 bp), which would in the com-\\nplete genome account for three million nucleotides. As a\\nFigure 4. Schematic summary of the various forms of variations which occur in genes. These might involve only one or a few base pairs (small\\nmutations) or large genomic regions (large mutations). Adapted from [9].\\n106\\nThe Human Genome: An Introduction\\ncomparison, the DNA sequence of a human and a chimpanzee\\nis estimated to differ 2% (1 in 50 base pairs).\\nSEARCHING FOR DISEASE GENES USING VARIATIONS\\nIN THE GENOME\\nThere is major interest among scientists in studying\\nvariations in genes, especially in the regulatory and protein-\\ncoding sequences, because such variations might be directly\\nrelated to specific diseases or other specific characteristics\\n(e.g., eye or hair color). The investigation of potential rela-\\ntionships of variations in specific genes with a specific dis-\\norder might be very useful if the candidate gene(s) to be\\ninvestigated can be well chosen. Such choice of candidate\\ngenes could be based on scientific knowledge or on new exper-\\nimental evidence (e.g., altered serum level of a protein in a spe-\\ncific patient group, microarray expression experiments, etc.).\\nGood candidate genes are, unfortunately, not always avail-\\nable. Therefore, genetic approaches have been developed in\\nthe past based on the analysis of highly polymorphic dinu-\\ncleotide repeat markers (microsatellites) in DNA samples from\\nindividuals from large families with multiple disease-affected\\nindividuals. The strategy is based on the identification of chro-\\nmosomal markers cosegregating with the disease in the fami-\\nlies. Such linkage studies are very attractive because they\\nallow identification of a chromosomal region on the genetic\\nmap which contains a disease-causing gene without requiring\\nany functional knowledge of the disease gene. Once a chro-\\nmosomal region with significant linkage is found, the disease-\\ncausing gene needs to be cloned and the responsible\\nmutation(s) identified. This positional cloning strategy has\\nbeen very successful, especially for identifying genes involved\\nin single-gene disorders (also called simple genetic disorders,\\nor Mendelian inherited disorders). Indeed even for some more\\ncommon disorders such as breast cancer or Alzheimer’s dis-\\nease, a positional cloning strategy has been successfully\\napplied and has led to the identification of the genes involved\\n(BRCA1 and BRCA2 in familial breast cancer, and presenilin\\ngenes in early onset Alzheimer’s disease, respectively).\\nAlthough it is clear that genetic tests for these mutations might\\nbe extremely useful for predicting disease risk in other mem-\\nbers of these families, it should be noted that defects in these\\ngenes can explain only a small fraction (usually less than 5%)\\nof the whole population of patients suffering these common\\ndisorders, consisting mainly of non-familial cases.\\nUnfortunately, however, the resolution that is obtained\\nusing these family studies is rather limited—at the very best,\\nup to a region of about one million bp. As this is still a very\\nlarge region—and may eventually contain more than 50\\ngenes—it is key to refining this region of interest. Because of\\ntheir high frequency in the genome, the analysis of SNP mark-\\ners has been proposed as a possible tool. SNP markers in the\\nregion of interest can be analyzed in a population of affected\\nindividuals and a population of matched healthy controls. For\\neach of the analyzed SNPs, the allele frequency in both pop-\\nulations is then compared. A statistically significant differ-\\nence in allele frequency of a genetic marker is suggestive for\\nan association of this marker with the disease.\\nFollowing the successes in genetic mapping and identifi-\\ncation of the molecular basis of Mendelian traits, attention\\nhas rapidly shifted to more complex and more prevalent\\ngenetic disorders that involve multiple genes and environ-\\nmental effects (e.g., cardiovascular disease, diabetes, and\\nschizophrenia). It is believed that SNPs could probably be\\nthe best available markers in the search for the origins of\\ncomplex genetic diseases. Moreover, it has been hypothe-\\nsized that ultimately, if enough SNP markers would become\\navailable with a chromosomal localization evenly dispersed\\nover the whole human genome, it should be feasible to\\ndirectly perform population-based whole-genome associa-\\ntion studies which would permit skipping of the initial step of\\nfamily-based linkage studies. As a consequence of the great\\npromise of SNPs, ten of the world’s pharmaceutical giants,\\nalong with five academic partners, entered into a close col-\\nlaboration in April 1999 called “The SNP Consortium.” The\\nmajor mission of this consortium is to create a high-quality,\\ndense, genome-wide SNP map, which will be made available\\nto the public. More specifically, The SNP Consortium aims\\nto generate genome SNP maps which would allow whole-\\ngenome, population-based association studies. It is estimated\\nthat this will require at least one marker every 5 to 50 kbp of\\nDNA. To cover the whole genome at this resolution would\\nrequire the identification and chromosomal localization of\\n200,000-300,000 new SNP markers. In July 2000, already\\nmore than 800,000 SNPs were made available to the public\\n(www.ncbi.nlm.nih.gov/snp; http://snp.cshl.org).\\nCOMPARATIVE GENOMICS—ANOTHER TOOL\\nFOR IDENTIFYING AND UNDERSTANDING GENES\\nRELEVANT IN HUMAN DISEASE\\nA powerful tool for understanding the human genome is\\ncomparison with the genome information from other organ-\\nisms; this area of research is called “comparative genomics”\\n[3]. The currently available sequence technology allows deter-\\nmination of the complete genome sequence of organisms\\nwithin reasonable time frames. In 1995, the first entire sequence\\nof an organism, Haemophilus influenza (1.8 Mbp), was pub-\\nlished. Since then, the complete genome sequence of a con-\\nstantly growing list of microorganisms (bacteria and viruses)\\nbecame known (size usually between 0.5-5 Mbp). The\\nsequence information can be used to identify specific genes and\\ntheir structure, regulation, and function, which might poten-\\ntially lead to new drugs which target a specific microorganism.\\nAerssens, Armstrong, Gilissen et al.\\n107\\nSaccharomyces cereviseae (baker’s yeast) was the first\\neukaryotic organism from which the entire genome sequence\\n(15 Mbp) was published (http://genome-www.stanford.edu/\\nSaccharomyces/). An enormous amount of information is\\nknown about the structure, regulation, and function of yeast\\ngenes. Of particular interest for developmental biology\\nresearch is the availability of the complete genome sequence\\nof the long roundworm Caenorhabditis elegans (97 Mbp);\\nthis animal consists of 959 somatic cells, the exact lineage of\\nwhich is known for every cell (http://elegans.swmed.edu/).\\nA cross-comparison of the complete gene sets of S. cerevisiae\\n(6,000 genes) and C. elegans (19,000 genes) has revealed\\nthat 23% of the proteins encoded by yeast genes have appar-\\nent homologues in the nematode worm, reflecting functions\\ncommon to both organisms [4].\\nThe fruit fly (Drosophila melanogaster) (137 Mbp) has a\\nlong history in genetic research, especially for its ease of cor-\\nrelation of genotype and phenotype. Because crucially impor-\\ntant gene functions and developmental processes appear to be\\nhighly conserved between species, the relevance for human\\ndisease research becomes clear. Moreover, there is also an\\nimportant conservation in the area of the cell-cycle control\\ngenes (and DNA repair and apoptosis), with immediate \\nrelevance to human cancer (http://flybase.bio.indiana.edu/).\\nThe mouse genome (3,000 Mbp) shows large subchromo-\\nsomal areas with a strong conservation of linkage (synteny)\\nbetween mouse and humans. This implies that, based on the\\nchromosomal localization of a gene on the mouse genome,\\npredictions can be made on the chromosomal localization of its\\nhuman homologue. Nearly every human gene appears to\\nhave a mouse homologue (http://www.ncbi.nlm.nih.gov/\\nHomology/). Because of their small body size, the short gen-\\neration time, and the technical ability to modify the DNA con-\\ntent of mice cells at the germline level, these animals provide\\nalso a powerful tool for studying gene expression and function\\nand for creating models of human disease (eventually by\\nmeans of knock-out and/or transgenic mice) [5].\\nAn interesting observation emerging from comparing\\ncomplete gene sets in model organisms known to date is that\\ngene number is not necessarily a good measure of complex-\\nity. For example, the fruit fly would be considered more\\nanatomically complex (with 10× more cells) than the nema-\\ntode, and the fruit fly undergoes a more complex develop-\\nmental process than C. elegans. Yet the fruit fly genome\\ncontains only 13,000 genes, compared with the 19,000 genes\\nfound in the nematode genome. It is generally expected that\\ncomparative genomic assessments will become increas-\\ningly important because they allow expansion of the utility\\nof the genomic information known and documented (“anno-\\ntated”) in one species toward other species, including\\nhumans.\\nGENOTYPE AND PHENOTYPE\\nThe two copies of a specific gene inherited from the\\nfather and the mother are not always identical, because for\\nmost genes many variants (alleles) exist. The combination of\\nthe two alleles present on the two homologous chromosomes\\nof the DNA is defined as the “genotype” for a specific genetic\\nvariation. For example, imagine an SNP in a gene with two\\npossible alternative alleles: allele A and allele G. The possible\\ngenotypes are thus A/A, A/G, and G/G. When two different\\nalleles of a gene are identified on the two homologous chro-\\nmosomes of an individual, the genotype is called “heterozy-\\ngous” (e.g., A/G); if the same allele is present on both\\nhomologous chromosomes, the genotype is “homozygous”\\n(e.g., A/A and G/G). More generally, the genotype of an indi-\\nvidual can be defined as the complete composition of an\\nindividual’s genome (including all the information on the\\nvariations within his/her genome), as has been defined at\\nconception. When used in clinical genetic applications,\\nhowever, the term genotype usually refers to some specific\\nvariation(s) in a small part of the DNA, often named for the\\ngene involved. For example, the APOE gene in the DNA can\\nexist as allele e2, e3, or e4, the latter of which is associated\\nwith an increased risk of developing Alzheimer’s disease. The\\nage of onset of the disease is lower in individuals with a geno-\\ntype harboring one or more copies of the e4 allele, namely the\\ne2/e4 or e3/e4, and especially the e4/e4 genotype.\\nOpposite to the genotype is the “phenotype,” which can\\nbe defined as the combination of all the observable or mea-\\nsurable characteristics of an individual (e.g., eye color, hair\\nstyle, body height, affected by disease, etc.). The phenotype\\nis—at least partially—determined by the genotype, because it\\ndepends on the level at which specific genes can be expressed.\\nThe latter depends on the variations in the DNA sequence but\\nalso on the environmental influences (e.g., nutrition status).\\nVery importantly, it should be pointed out that although\\nthe phenotype may appear to be equal in two individuals, their\\ngenotypes might be different. This could be due to a significant\\nenvironmental influence which overrules the genetic impreg-\\nnation, or alternatively because some of the possible genotypes\\ndo not result in phenotypic differences (e.g., genotypes A/A\\nand A/G can both have straight hair, while only genotype G/G\\nshows curly hair—Fig. 1). In a molecular diagnostic setting,\\nthe determination of the genotype usually aims to predict the\\nphenotype. Indeed the genotype might sometimes be fully pre-\\ndictable (especially in single-gene Mendelian inherited disor-\\nders, such as cystic fibrosis). Unfortunately, most often the\\nanalyzed genotype is not fully predictable for the phenotype\\nbut merely allows assignment of a certain risk level to an indi-\\nvidual for expressing or developing a specific phenotype. For\\nexample, susceptibility-conferring genotypes at the BRCA1\\nand BRCA2 gene loci confer a relative risk of breast cancer of\\n108\\nThe Human Genome: An Introduction\\nabout 5. Consequently, it is strongly advised that all results of\\ngenetic testing are accompanied by an interpretation for each\\nmolecular genetic diagnostic report to be used in the clinic.\\nGENETICS AND GENOMICS IN CANCER\\nAs mentioned above, familial cases of some specific can-\\ncer types are known, indicative of an inherited trait similar to\\nany other genetic disorder. For some of these cancer types,\\nsuccessful positional cloning projects have allowed identifi-\\ncation of genes harboring mutations which cause the disease.\\nAs these mutations are inherited by the next generation, it\\nimplies that the responsible gene defect is present in the\\ngermline cells. At present, more than 20 different hereditary\\ncancer syndromes have been defined and attributed to specific\\ngermline mutations. Collectively, these syndromes affect\\napproximately 1% of all cancer patients [6]. For several of the\\ninherited cancer syndromes, genetic testing for disease suscep-\\ntibility is feasible and already part of the clinical management\\nof affected families. Controversy on its value has been raised,\\nhowever, especially in cases where the risks of developing\\ncancer associated with a predisposing mutation are less cer-\\ntain, or where there is no effective intervention to offer those\\nwith a positive result [7].\\nMost cancer patients do not have any pronounced fam-\\nily history, yet genomic defects are at the basis of the dis-\\nease. The genetic information present in normal cells can\\nalso be altered (e.g., due to incorrect DNA duplication dur-\\ning cell division), either by gross chromosomal changes\\nsuch as translocations, deletions, inversions, and amplifica-\\ntions, or through more subtle changes such as point muta-\\ntions and microdeletions [8]. The accumulation of these\\ngenetic alterations can finally lead to the expression of the\\nfull cancer phenotype. It should be noticed that—apart from\\nthe familial cases—these changes in the DNA do occur in\\nthe somatic cells and do not transfer to the germline cells.\\nConsequently, these abnormalities are not inherited by the\\nchildren of these patients.\\nHistorically, chromosomal abnormalities in tumors were\\nfirst recognized when an unusually small chromosome, the\\n“Philadelphia chromosome,” was observed in white blood\\ncells as a hallmark of chronic myeloid leukemia. The signifi-\\ncance of these chromosomal abnormalities has only relatively\\nrecently become clear by a combination of improved cytoge-\\nnetics and molecular biology. The central concept is that of\\nproto-oncogenes and tumor suppressor genes: normal cellular\\ngenes controlling growth, development, differentiation, DNA\\nrepair, and DNA modification become deregulated in the neo-\\nplastic cancer cell due to mutations, fusions, or deletions. The\\nnormal structure of a resident proto-oncogene may be con-\\nverted to a dominant oncogene by mutations or chromosomal\\nrearrangements. Such conversion in one copy of the gene (one\\nchromosome homologue) is sufficient to result in neoplastic\\ntransformation. On the contrary, loss or inactivation of tumor\\nsuppressor genes may release a cell from constraints imposed\\nby these genes, resulting in uncontrolled growth. Their behav-\\nior is recessive, and both allele copies must be lost for tumor\\nactivation to occur. Therefore, recurrent deletions of chromo-\\nsomal material are recognized as indications for the presence\\nof tumor-suppressor genes. On the other hand, recurring spe-\\ncific chromosomal aberrations (translocations, amplifications,\\nand inversions) have been instrumental in identifying proto-\\noncogenes. The cloning of the chromosomal breakpoints of\\nsuch aberrations has proven to be an effective strategy for\\nidentifying mutant genes in tumors (e.g., ETV6 in leukemia,\\nc-MYC in Burkitt’s lymphoma). At present, more than 50\\nchromosomal translocation breakpoints have been molecu-\\nlarly cloned and the involved genes identified. The vast\\nmajority of these tumors were of hematopoietic origin, as\\ncytogenetic data on these tumors are easier to obtain and are\\nthus more extensively studied. It is clear that cytogenetic\\nanalysis might allow subtyping of patients with an apparently\\nsimilar phenotype. Depending on the chromosomal abnor-\\nmalities (and thus the genes involved), the efficacy of therapy\\ncan be predicted to a certain extent. Therefore, cytogenetic\\nanalysis has now become a routine analysis in many centers\\n(http://www.waisman.wisc.edu/cytogenetics/Bmproject/\\nCancerCyto.htmlx). Finally, the colorful fluorescence in situ\\nhybridization (FISH) technique allows direct identification\\nof the breakpoint region involved in a chromosomal abnor-\\nmality at a relative high resolution (10-100 kbp). Specific\\nprobes to be used by FISH which recognize recurrent abnor-\\nmalities of specific regions of the genome are now commer-\\ncially available for routine analysis of cancer cells derived\\nfrom oncology patients.\\nA nice overview on the currently known genetics and\\ngenomics behind cancers is provided in a subsection of the\\nweb site of the National Center for Biotechnology\\nInformation (NCBI), which specifically deals with this\\ntopic (http://www.ncbi.nlm.nih.gov/disease/Cancer.html).\\nOngoing research in oncology is further directed toward a\\nmore complete and basic understanding of why some\\nsomatic cells at a certain point in time become tumor cells.\\nIn this respect, the National Cancer Institute coordinates the\\nCancer Genome Anatomy Project (CGAP), which provides\\na valuable resource of information and technological tools\\nrequired to analyze the molecular anatomy of the cancer\\ncell (http://www.ncbi.nlm.nih.gov/ncicgap).\\nTOWARD GENETICS AND GENOMICS APPLICATIONS\\nIN THE CLINIC\\nOriginally, molecular genetics was used in medicine only\\nto identify gene defects in major single-gene disorders such as\\nAerssens, Armstrong, Gilissen et al.\\n109\\ncystic fibrosis. The excitement in the field has shifted gradu-\\nally toward more common and complex genetic disorders. It is\\ntherefore not surprising that the amount of genetic information\\non an ever-increasing number of diseases has exploded over\\nthe past few years. The Online Mendelian Inheritance In Man\\n(OMIM) is a database of bibliographic information about\\nhuman genes and genetic disorders and is freely available\\nonline (http://www.ncbi.nlm.nih.giv/omim/). With more than\\n10,000 entries (newly defined for each distinct disease gene or\\ngenetic disorder for which sufficient information exists), it\\nprovides probably the most comprehensive, authoritative, and\\ntimely compendium of information in human genetics.\\nClinicians can use OMIM as an aid in differential diagnosis by\\nsearching the database using key clinical features of a patient.\\nAn important question for the clinician is which impact on\\nfuture medical practice might be expected from ongoing\\nresearch activities in genomics and genetics. The main contri-\\nbution to date has been in the identification of new molecular\\ntargets for drug action, which might in the long term result in\\nnew and better drug therapies. It is expected, however, that\\nclinical practice will also be increasingly affected by new diag-\\nnostic tests based on genetic markers associated with increased\\ndisease risk, therapeutic efficacy, or adverse events. In this\\nrespect, the research area designated as pharmacogenomics is\\nexpected to become a driving force toward a more rational use\\nof pharmaceutical products. Expensive therapies might possi-\\nbly no longer be authorized without a definite diagnosis based\\non a genetic test. Validated genetic tests enabling prediction of\\nincreased risk on disease development might eventually lead to\\na shift from curative toward predictive treatment, long before\\nclinical symptoms of the disease can be observed.\\nIn conclusion, it might be expected that genomics and\\ngenetics will largely impact future medical practice. Several\\ngenomics-based applications are on their way to enter the\\nclinic within the next few years, and many more will most\\nprobably follow in a later stage. For clinicians of the 21st\\ncentury, it will be key to be well prepared and open-minded\\nfor this molecular future of medicine.\\nREFERENCES\\n1 Macilwain C. World leaders heap praise on human genome\\nlandmark. Nature 2000;405:983-984.\\n2 Lockhart DJ, Winzeler EA. Genomics, gene expression and\\nDNA arrays. Nature 2000;405:827-836.\\n3 Bentley DR. Decoding the human genome sequence. Hum\\nMol Genet 2000;9:2353-2358.\\n4 Rubin GM, Yandell MD, Wortman JR et al. Comparative\\ngenomics of the eukaryotes. Science 2000;87:2204-2215.\\n5 O’Brien S, Menotti-Raymond M, Murphy WJ et al. The\\npromise of comparative genomics in mammals. Science\\n1999;286:458-481.\\n6 Fearon ER. Human cancer syndromes: clues to the origin and\\nnature of cancer. Science 1997;278:1043-1050.\\n7 Ponder B. Genetic testing for cancer risk. Science\\n1997;278:1050-1054.\\n8 Lengauer C, Kinzler KW, Vogelstein B. Genetic instabilities\\nin human cancers. Nature 1998;396:643-649.\\n9 Varmus H, Weinberg RA. Genes and the biology of cancer.\\nNew York: Scientific American Library, 1993:10-14.\\nADDITIONAL READING\\nBrown PO, Hartwell L. Genomics and human disease—varia-\\ntions on variation. Nat Genet 1998;18:91-93.\\nCollins FS, Guyer MS, Chakravarti A. Variations on a theme:\\ncataloging human DNA sequence variation. Science\\n1997;278:1580-1581.\\nCollins FS. Medical and societal consequences of the human\\ngenome project. N Engl J Med 1999;341:28-37.\\nHamosh A, Scott AE, Amberger J et al. Online Mendelian\\nInheritance in Man (OMIM). Hum Mutat 2000;15:57-61.\\nHoltzman NA, Marteau TM. Will genetics revolutionize medi-\\ncine? N Engl J Med 2000;343:141-144.\\nLander ES, Schork NJ. Genetic dissection of complex traits.\\nScience 1994;265:2037-2048.\\nPoste G. Molecular medicine and information-based targeted\\nhealthcare. Nat Biotech 1998;16(suppl 1):19-21.\\nRoses AD. Pharmacogenetics and future drug development and\\ndelivery. The Lancet 2000;355:1358-1361.\\nSchafer AJ, Hawkins JR. DNA variation and the future of human\\ngenetics. Nat Biotech 1998;16:33-39.\\nStrachan T, Read AP. Human Molecular Genetics, 2nd Ed. Oxford:\\nBios Scientific Publishers Ltd., 1999:1-53, 139-168, 295-314,\\n351-375, 427-444.\\nWolf CR, Smith G, Smith RL. Pharmacogenetics. BMJ\\n2000;320:987-990.\\nView publication stats\\n'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9ace3545-a9cf-4d53-ad80-e60deb145737",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Question: What is the main topic of the document?\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'question' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[10], line 7\u001b[0m\n\u001b[0;32m      3\u001b[0m user_question \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mWhat is the main topic of the document?\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m      4\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mQuestion: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00muser_question\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m----> 7\u001b[0m inputs \u001b[38;5;241m=\u001b[39m tokenizer\u001b[38;5;241m.\u001b[39mencode(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mquestion: \u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m question \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m context: \u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m context, return_tensors\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpt\u001b[39m\u001b[38;5;124m\"\u001b[39m, max_length\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m512\u001b[39m, truncation\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m      8\u001b[0m outputs \u001b[38;5;241m=\u001b[39m model\u001b[38;5;241m.\u001b[39mgenerate(inputs)\n\u001b[0;32m      9\u001b[0m answer \u001b[38;5;241m=\u001b[39m tokenizer\u001b[38;5;241m.\u001b[39mdecode(outputs[\u001b[38;5;241m0\u001b[39m], skip_special_tokens\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "\u001b[1;31mNameError\u001b[0m: name 'question' is not defined"
     ]
    }
   ],
   "source": [
    "# Define your query\n",
    "\n",
    "user_question = \"What is the main topic of the document?\"\n",
    "print(f\"Question: {user_question}\")\n",
    "\n",
    "\n",
    "inputs = tokenizer.encode(\"question: \" + question + \" context: \" + context, return_tensors=\"pt\", max_length=512, truncation=True)\n",
    "outputs = model.generate(inputs)\n",
    "answer = tokenizer.decode(outputs[0], skip_special_tokens=True)\n",
    "print(f\"Answer: {answer}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8c24464-a46c-4146-b640-3da7b7873aff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
